kernel/frame_allocator/src/lib.rs

```diff
@@ -421,120 +697,48 @@ impl AllocatedFrames {
-        } else {
-            (&FREE_GENERAL_FRAMES_LIST, MemoryRegionType::Free)
-        };
-        // trace!("frame_allocator: deallocating {:?}, typ {:?}", self, typ);
-
-        // Simply add the newly-deallocated chunk to the free frames list.
-        let mut locked_list = list.lock();
-        let res = locked_list.insert(Chunk {
-            typ,
-            frames: self.frames.clone(),
-        });
-        match res {
-            Ok(_inserted_free_chunk) => (),
-            Err(c) => error!("BUG: couldn't insert deallocated chunk {:?} into free frame list", c),
-        }
-        
-        // Here, we could optionally use above `_inserted_free_chunk` to merge the adjacent (contiguous) chunks
-        // before or after the newly-inserted free chunk. 
-        // However, there's no *need* to do so until we actually run out of address space or until 
-        // a requested address is in a chunk that needs to be merged.
-        // Thus, for performance, we save that for those future situations.
+impl<const S: MemoryState> Ord for Frames<S> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.frames.start().cmp(other.frames.start())
     }
 }
-
-impl<'f> IntoIterator for &'f AllocatedFrames {
-    type IntoIter = AllocatedFramesIter<'f>;
-    type Item = AllocatedFrame<'f>;
-    fn into_iter(self) -> Self::IntoIter {
-        AllocatedFramesIter {
-            _owner: self,
-            range_iter: self.frames.iter(),
-        }
+impl<const S: MemoryState> PartialOrd for Frames<S> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
     }
 }
-
-/// An iterator over each [`AllocatedFrame`] in a range of [`AllocatedFrames`].
-///
-/// We must implement our own iterator type here in order to tie the lifetime `'f`
-/// of a returned `AllocatedFrame<'f>` type to the lifetime of its containing `AllocatedFrames`.
-/// This is because the underlying type of `AllocatedFrames` is a [`FrameRange`],
-/// which itself is a [`RangeInclusive`] of [`Frame`]s.
-/// Currently, the [`RangeInclusiveIterator`] type creates a clone of the original
-/// [`RangeInclusive`] instances rather than borrowing a reference to it.
-///
-/// [`RangeInclusive`]: range_inclusive::RangeInclusive
-pub struct AllocatedFramesIter<'f> {
-    _owner: &'f AllocatedFrames,
-    range_iter: RangeInclusiveIterator<Frame>,
+impl<const S: MemoryState> PartialEq for Frames<S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.frames.start() == other.frames.start()
+    }
 }
-impl<'f> Iterator for AllocatedFramesIter<'f> {
-    type Item = AllocatedFrame<'f>;
-    fn next(&mut self) -> Option<Self::Item> {
-        self.range_iter.next().map(|frame|
-            AllocatedFrame {
-                frame, _phantom: PhantomData,
-            }
-        )
+impl<const S: MemoryState> Borrow<Frame> for &'_ Frames<S> {
+    fn borrow(&self) -> &Frame {
+        self.frames.start()
```

Comment:
yes, this is necessary because when we create a cursor for the `RBTree`, we search using a `Frame` value, and so we need `Frames` to implement `Borrow<Frame>`

Reply:
added a disclaimer